### ISSUE TITLE
Need more intuitive method to extend Model `$fields`

### ISSUE THREAD

**Issue by justlevine** [wp-graphql/wp-graphql#2198]
# Tl;dr
Currently, adding a field to an existing GraphQL model is bad DX, requiring two WP Filters: first to set the field to null, and then to resolve the value. 

The fix is to pass the model $data to `graphql_return_model_data`.

# Problem

There's currently no intuitive method for adding a field to an existing WPGraphQL model.

[`graphql_return_model_data`](https://www.wpgraphql.com/filters/graphql_return_modeled_data/) in its current form is only really suitable for removing fields. Since `Model::$data` isnt passed to the filter, there is no way to resolve the new field to anything meaningful. 

Meanwhile, [`graphql_pre_return_field_from_model`](https://www.wpgraphql.com/filters/graphql_pre_return_field_from_model/) only handles fields that have already been registered.

As such, the only way to currently handle adding fields to a Model is to use _both_ filters, first setting the field to `null` in `graphql_return_model_data, and _then_ resolving it in `graphql_pre_return_field_from_model`:

```php
add_filter(
  'graphql_return_modeled_data',
  function( array $fields, string $model_name ) {
    if ( 'myModel' === $model_name ) {
      $fields['myCustomField'] = null;
    }
    return $fields;
  },
  10,
  2
);

add_filter(
  'graphql_pre_return_field_from_model'
  function( $return_field, string $key, string $model_name, $data ) {
    if ( 'myModel' === $model_name && 'myCustomField' === $key ) {
      $return_field = get_my_custom_field_value( $data );
    }
    return $return_field;
  },
  10,
  4
);
```

# Recommended Solution

The ideal solution (imo) would be to simply add the model `$data` to [`graphql_return_modeled_data`.

This is easy enough to do without breaking back-compat, by just appending `$this->data` as the last arg [in the filter](https://github.com/wp-graphql/wp-graphql/blob/194e89306ce0ebd158887f92b3b035dfc5a14972/src/Model/Model.php#L486).

__However__, from a DX standpoint, $data should really be earlier in the list of args:

```php
$this->fields = apply_filters( 'graphql_return_modeled_data', $this->fields, $this->get_model_name(), $this->data, $this->visibility, $this->owner, $this->current_user );
```

That would however constitute a breaking change.

**Comment by justlevine** [comment]
@jasonbahl I'm happy to submit a PR, just let me know if I should go with the breaking or non-breaking version.

**Comment by jasonbahl** [comment]
@justlevine I would prefer a non-breaking way. 

One thing we could consider is adding another filter, and deprecating the old one, but still letting it exist. 

Something like: 

```
apply_filters( 'graphql_model_prepare_fields', $this->fields, $this->get_model_name(), $this->data, $this->visibility, $this->owner, $this->current_user );
```

Then, add a deprecated.php file and something like: 

```php
add_filter( 'graphql_model_prepare_fields', function( $fields, $model_name, $data, $visibility, $owner, $current_user ) {
  // deprecated filter
  return apply_filters( 'graphql_return_modeled_data', $fields, $model_name, $visibility, $owner, $current_user, $data );
}, 10, );
```

Another option we could consider is passing the Model and allowing the filter access to the properties. 

So something like: 

```php
apply_filters( 'graphql_model_prepare_fields', $this->fields, $this );
```

Then in the deprecated filter we could do: 

```php
add_filter( 'graphql_model_prepare_fields', function( $fields, $model ) {
  // deprecated filter, but executes for back compat
  return apply_filters( 'graphql_return_modeled_data', $fields, $model->get_model_name(), $model->visibility, $model->owner, $model->current_user, $model->data );
}, 10, );
```
